e2etest/clean/ec2/clean_ec2.go

```diff
@@ -0,0 +1,71 @@
+package main
+
+import (
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"log"
+	"time"
+)
+
+func main() {
+	// set up aws go sdk ec2 client
+	testSession, err := session.NewSession()
+	if err != nil {
+		log.Fatalf("Error creating session %v", err)
+	}
+	ec2client := ec2.New(testSession)
+
+	// Get list of instance
+	//State filter
+	instanceStateName := "instance-state-name"
+	instanceStateValue := "running"
+	instanceStateValues := []*string{&instanceStateValue}
+	instanceStateFilter := ec2.Filter{Name: &instanceStateName, Values: instanceStateValues}
+
+	//Tag filter
+	instanceTagName := "tag:Name"
+	instanceTagValueIntegTestSampleApp := "Integ-test-Sample-App"
+	instanceTagValueIntegTestAOC := "Integ-test-aoc"
+	instanceTagValues := []*string{&instanceTagValueIntegTestSampleApp, &instanceTagValueIntegTestAOC}
+	instanceTagFilter := ec2.Filter{Name: &instanceTagName, Values: instanceTagValues}
+
+	//get instances to delete
+	var deleteInstances []ec2.Instance
+	var nextToken *string
+	for {
+		describeInstancesInput := ec2.DescribeInstancesInput{Filters: []*ec2.Filter{&instanceStateFilter, &instanceTagFilter}, NextToken: nextToken}
+		describeInstancesOutput, err := ec2client.DescribeInstances(&describeInstancesInput)
+		if err != nil {
+			log.Fatalf("Failed to get instance for error %v", err)
+		}
+		for _, reservation := range describeInstancesOutput.Reservations {
+			for _, instance := range reservation.Instances {
+				//only delete instance if older than 30 days
+				if time.Now().UTC().Add(-1 * time.Hour * 24 * 30).After(*instance.LaunchTime) {
+					deleteInstances = append(deleteInstances, *instance)
+				}
+			}
+		}
+		if describeInstancesOutput.NextToken == nil {
+			break
+		}
+		nextToken = describeInstancesOutput.NextToken
+	}
+
+	if len(deleteInstances) < 1 {
+		log.Print("No instances to delete")
+		return
+	}
+
+	//delete instances
+	var instanceIds []*string
+	for _, instance := range deleteInstances {
+		instanceIds = append(instanceIds, instance.InstanceId)
+		log.Printf("Delete instance %s tags %s launch-date %s", *instance.InstanceId, instance.Tags, instance.LaunchTime)
+	}
+	terminateInstancesInput := ec2.TerminateInstancesInput{InstanceIds: instanceIds}
+	_, err = ec2client.TerminateInstances(&terminateInstancesInput)
+	if err != nil {
+		log.Fatalf("Failed to terminate instance %v", err)
```

Comment:
What would cause this to fail? Should this be fatal? Would it make sense to continue and try to terminate the other instances?

Reply:
See https://docs.aws.amazon.com/goto/WebAPI/ec2-2016-11-15/TerminateInstances
This can fail if an instance is protected. We should not protect these instances so this should be fine. Instances within the same az will not terminate but instances outside that az will. I think this is fine.